src/Application/Kernel.php

```diff
@@ -161,4 +149,57 @@ private function initDiscovery(Container $container): void
             $discovery->storeCache();
         }
     }
+
+    private function discoverInstalledPackageLocations(): void
+    {
+        $composerPath = path($this->root, 'vendor/composer');
+        $installed = path($composerPath, 'installed.json');
+
+        if (is_file($installed)) {
+            $installed = $this->loadJsonFile($installed);
```

Comment:
So here, I'd be curious what something to the effect of the following would give you. 🙂 

```php
$installedContents = file_get_contents($installed);

// Splits each package section {} into it's own block
preg_match_all('/(?<=\s{8})\{[\s\S]+?\n\s{8}\}/', $installedContents, $packages);

foreach ($packages as $package) {
    // Decodes for the specific package
    $package = json_decode($package, true);
}
```

Reply:
i'll give it a shot and see what happens.

Reply:
## Results

Also added `laravel/framework` to the Tempest `composer.json` (HOLY CRAP Symfony, are you entirely built on polyfills???)

Using the regex:
```php
if (is_file($installed)) {
    $installed = file_get_contents($installed);
    preg_match_all('~(?<=\s{8})\{[\s\S]+?\n\s{8}}~', $installed, $packages);

    foreach ($packages[0] as $package) {
        $package = json_decode($package, true);
        $packagePath = realpath(path($composerPath, $package['install-path']));
```

Benchmarks for regex and my version with Laravel included...

My version:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done

0.22264194488525
0.22401404380798
0.27412104606628
0.24248886108398
0.22104001045227
0.26659607887268
0.20401000976562
0.2078320980072
0.18999195098877
0.23425197601318

(Avg: 0.228 sec)
```

Regex:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done

0.27564597129822
0.25507497787476
0.26790499687195
0.21789503097534
0.20109605789185
0.19769716262817
0.25411415100098
0.21002793312073
0.22865700721741
0.25373101234436

(Avg: 0.236 sec)
```

Roughly speaking, very little performance change. By the way, that's ~100 total package entries in `installed.json` so i imagine it gives a better than average estimation. 😃

Reply:
i've seen references to more advanced JSON handlers that might help performance, but haven't worked with any of them previously, so maybe it's something we can try in a later revision. Let me know whether to update this PR to the regex or not.

Reply:
Okay, so I am getting very different numbers from you. I started the evening genuinely curious and confused and ended the evening over-optimizing microseconds. 😅 

For some reason I cannot replicate your timings at all. I don't know if it is a PHP version mismatch (I am on 8.3) or if you are running on a slower disk (HDD?). See below for the timing I am currently getting and the "optimized" code I am currently using.


```bash
# json_decode
0.0014290809631348n
0.0011060237884521n
0.0010828971862793n
0.0011448860168457n
0.0010919570922852n
0.0010881423950195n
0.0010750293731689n
0.0010759830474854n
0.0011980533599854n
0.0010321140289307n

# RegEx
0.00080204010009766n
0.000762939453125n
0.00078010559082031n
0.00077986717224121n
0.00078606605529785n
0.00073504447937012n
0.0007779598236084n
0.000762939453125n
0.00078701972961426n
0.00080013275146484n
```

```php
private function discoverInstalledPackageLocations(): void
{
    $startup = microtime(true);

    $composerPath = path($this->root, 'vendor/composer');
    $installed = path($composerPath, 'installed.json');

    if (! is_file($installed)) {
        return;
    }

    $content = file_get_contents($installed);
    $packages = json_decode($content, true)['packages'];

    foreach ($packages as $package) {
        $packagePath = realpath(path($composerPath, $package['install-path']));

        if (
            ! isset($package['require']['tempest/framework']) ||
            empty($package['autoload']['psr-4'])
        ) {
            continue;
        }

        foreach ($package['autoload']['psr-4'] as $namespace => $namespacePath) {
            $namespacePath = path($packagePath, $namespacePath);
            $this->addDiscoveryLocation($namespace, $namespacePath);
        }
    }

    die(microtime(true) - $startup);
}
```

```php
private function discoverInstalledPackageLocations(): void
{
    $startup = microtime(true);

    $composerPath = path($this->root, 'vendor/composer');
    $installed = path($composerPath, 'installed.json');

    if (! is_file($installed)) {
        return;
    }

    $content = file_get_contents($installed);

    preg_match_all('/\s{8}\{.*?tempest\/framework.*?\n\s{8}\}/s', $content, $packages);

    foreach ($packages[0] as $package) {
        $package = json_decode($package, true);
        $packagePath = realpath(path($composerPath, $package['install-path']));

        if (
            ! isset($package['require']['tempest/framework']) ||
            empty($package['autoload']['psr-4'])
        ) {
            continue;
        }

        foreach ($package['autoload']['psr-4'] as $namespace => $namespacePath) {
            $namespacePath = path($packagePath, $namespacePath);
            $this->addDiscoveryLocation($namespace, $namespacePath);
        }
    }

    die(microtime(true) - $startup);
}
```

Full disclosure, I did adjust my RegEx slightly so that now it is _only_ pulling the packages that mention Tempest in. It did not improve speed dramatically, but it is a difference from yours.

Reply:
I suspect the difference for regex probably has to do with differences in CPU. I still don't think regex is the way to go, as we'll probably miss edge cases leading to subtle bugs.

Reply:
Yeah, very weird. i'm running 8.3 through WSL. i disabled all extensions to the minimum that would run the script (mbstring and pdo)...

Your version with `json_decode()`:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done

0.21244215965271
0.19042801856995
0.19336795806885
0.19525384902954
0.179358959198
0.20081305503845
0.20244216918945
0.18885803222656
0.22774887084961
0.21510100364685
```

Your version with regex:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done
0.57280683517456
0.57812809944153
0.58618807792664
0.56667304039001
0.58222198486328
0.5756139755249
0.5791130065918
0.5681848526001
0.56858205795288
0.57621693611145
```
That difference from your numbers is _crazy_ ...

i'll put up results using 8.3 from PowerShell in a separate comment.

Reply:
@jtheuerkauf - Are you using WSL 2 and have the files on /mnt/c? If so, you are probably running into the slowness of the 9P protocol.

Reply:
Files are on a different mount, but otherwise normal. You may be right though. Here's 8.3 from PS:

`json_decode()`:
```
> for ($i = 0; $i -lt 10; $i++) { php .\tempest }

0.004680871963501
0.0046839714050293
0.0047390460968018
0.0050721168518066
0.0051999092102051
0.0046918392181396
0.0047049522399902
0.0047979354858398
0.0047011375427246
0.0047698020935059

Avg: 0.0048041582107544 sec
```


Regex:
```
> for ($i = 0; $i -lt 10; $i++) { php .\tempest }

0.00072097778320312
0.00059008598327637
0.00069689750671387
0.00060105323791504
0.00063681602478027
0.00059008598327637
0.00056695938110352
0.0006558895111084
0.00058507919311523
0.00063514709472656

Avg: 0.00062789916992187 sec
```

---

So, at the scale of thousandths of a second, the RegEx is faster. However, i lean toward @brendt's concern about reliable accuracy. But if the time difference is convincing enough, i'm open to switching to it.

My last thoughts on it: practically speaking, once caching is turned on, the time difference only matters once, right? Hell, discovery and caching the packages could be part of the composer post-install scripts, couldn't it? It never even has to occur for actual application requests/commands.

---

Anyway, how do we feel about this version? Any specific changes to make?

Reply:
I say we go with the non-regex approach. Sorry @aidan-casey 🫣

Regarding cache: local development won't have it enabled, so we need it to be performant enough still. I think the current numbers look good enough

Reply:
Vendor discovery can be cached locally btw, we're already planning on adding a discovery:cache command for that

Reply:
Hey, I get it. You guys don't care about speed. 😢🙃